Fixed HWIntrinsic reflection bug with `NextCallReturnAddress` intrinsic

TIHan opened the pull request:
Should resolve: https://github.com/dotnet/runtime/issues/71950

**Description**

@jakobbotsch and @kunalspathak helped me a bit with this. It's actually quite tricky.

An assert was getting hit in codegen as a block did not have its `bbEmitCookie` set. This was caused by a `LABEL` node not being associated with a `CALL`. Instead, there was a `HWINTRINSIC` node in its place. 

The `LABEL` node gets created in the importer when it handles an internal intrinsic `NI_System_StubHelpers_NextCallReturnAddress`. For this to be legal, the label must be followed by a `CALL`, but it's not. This must mean that the HWIntrinsic was expanded in the importer and we do not have a `CALL`.

Consider this example:
```csharp
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

var Op1ElementCount = Unsafe.SizeOf<Vector256<Double>>() / sizeof(Double);
var Op2ElementCount = Unsafe.SizeOf<Vector256<Double>>() / sizeof(Double);

var _data1 = new Double[Op1ElementCount];
var _data2 = new Double[Op2ElementCount];

unsafe
{
    fixed (Double* ptr = _data1)
    {
        fixed (Double* ptr2 = _data2)
        {
            var result = typeof(Avx).GetMethod(nameof(Avx.MaskStore), new Type[] { typeof(Double*), typeof(Vector256<Double>), typeof(Vector256<Double>) })
                 .Invoke(null, new object[] {
                            Pointer.Box(ptr, typeof(Double*)),
                            Avx.LoadVector256(ptr),
                            Avx.LoadVector256(ptr2)
                 });
        }
    }
    fixed (Double* ptr = _data1)
    {
        fixed (Double* ptr2 = _data2)
        {
            var result = typeof(Avx).GetMethod(nameof(Avx.MaskStore), new Type[] { typeof(Double*), typeof(Vector256<Double>), typeof(Vector256<Double>) })
                 .Invoke(null, new object[] {
                            Pointer.Box(ptr, typeof(Double*)),
                            Avx.LoadVector256(ptr),
                            Avx.LoadVector256(ptr2)
                 });
        }
    }
}
```

We are using reflection to get the method `Avx.MaskStore` and then we invoke it. 
We actually need to do this twice in order for the bug to get reproduced because the `MethodInvoker` in corelib has a first-time path:
```csharp
                if (!_invoked)
                {
                    // The first time, ignoring race conditions, use the slow path.
                    _invoked = true;
                }
                else
                {
                    if (RuntimeFeature.IsDynamicCodeCompiled)
                    {
                        _invokeFunc = InvokerEmitUtil.CreateInvokeDelegate(_method);
                    }
                    _strategyDetermined = true;
                }
```

Then you will see that `CreateInvokeDelegate` emits the intrinsic `NextCallReturnAddress`:
```csharp
            il.Emit(OpCodes.Call, Methods.NextCallReturnAddress()); // For CallStack reasons, don't inline target method.
            il.Emit(OpCodes.Pop);
```

The solution to this is to simply not expand the HWIntrinsic/Unsafe `CALL` if we detect a `NextCallReturnAddress` intrinsic in the importer.

**Acceptance Criteria**
- [x] x86 HWIntrinsic tests need to run in CI
- [x] CI Passes

TIHan committed:
Adding more asserts

TIHan committed:
Another assert

TIHan committed:
Force add a label

TIHan committed:
Extra check

jakobbotsch commented:
`GT_LABEL` represents the return address of an upcoming call instruction. It is used for pinvokes, tailcalls and some reflection invokes now (due to the use of `NextCallReturnAddress` intrinsic to inhibit inlining).

The assertion failure probably indicates that the JIT removed the upcoming call instruction (maybe it was an intrinsic that was force expanded?) so now we have a dangling `GT_LABEL` that will not refer to anything.

I think in this case it would be better to ensure codegen generated it as the value 0 rather than always inserting a label that will end up pointing to some random instruction.

TIHan commented:
> I think in this case it would be better to ensure codegen generated it as the value 0 rather than always inserting a label that will end up pointing to some random instruction.

Only inserted it to see how things would behave in CI.

I think the bug is happening in `emitRemoveJumpToNextInst` - if I comment `emitRemoveJumpToNextInst();` out, I do not get the assert.

TIHan committed:
Remove cookie set

TIHan commented:
Minimal repo:
```csharp
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

var Op1ElementCount = Unsafe.SizeOf<Vector256<Double>>() / sizeof(Double);
var Op2ElementCount = Unsafe.SizeOf<Vector256<Double>>() / sizeof(Double);

var _data1 = new Double[Op1ElementCount];
var _data2 = new Double[Op2ElementCount];

unsafe
{
    fixed (Double* ptr = _data1)
    {
        fixed (Double* ptr2 = _data2)
        {
            var result = typeof(Avx).GetMethod(nameof(Avx.MaskStore), new Type[] { typeof(Double*), typeof(Vector256<Double>), typeof(Vector256<Double>) })
                 .Invoke(null, new object[] {
                            Pointer.Box(ptr, typeof(Double*)),
                            Avx.LoadVector256(ptr),
                            Avx.LoadVector256(ptr2)
                 });
        }
    }
    fixed (Double* ptr = _data1)
    {
        fixed (Double* ptr2 = _data2)
        {
            var result = typeof(Avx).GetMethod(nameof(Avx.MaskStore), new Type[] { typeof(Double*), typeof(Vector256<Double>), typeof(Vector256<Double>) })
                 .Invoke(null, new object[] {
                            Pointer.Box(ptr, typeof(Double*)),
                            Avx.LoadVector256(ptr),
                            Avx.LoadVector256(ptr2)
                 });
        }
    }
}
```

jakobbotsch commented:
> Only inserted it to see how things would behave in CI.

After thinking about this some more it may not be super unreasonable as a fix for this edge case. I suppose the alternative would be to walk upcoming BBs and IR nodes in the codegen for `GT_LABEL` to scout for a `GT_CALL` that will be bound to, and then generate a zero if no `GT_CALL` is found. That's not all that pretty either.

TIHan committed:
Added check for compHasNextCallRetAddr for hw intrinsics

TIHan committed:
Added additional checks

TIHan commented:
@dotnet/jit-contrib This is ready for review.

It took a long time to get to this point, but thanks to @jakobbotsch for helping me understand the scenarios with `GT_LABEL` and the `NextCallReturnAddress`.

kunalspathak commented:
Did you get chance to see why this tests is not run or failed in our nightly/weekly CI pipelines?

TIHan commented:
> Did you get chance to see why this test is not ran or failed in our nightly/weekly CI pipelines?

I've still been trying to figure it out but I will look at it again on Monday.

jakobbotsch commented:
The test only fails in debug codegen which we don't see in CI. In checked/release I expect that the `GT_LABEL` gets removed as dead code since its result is unused.

TIHan commented:
> In checked/release I expect that the GT_LABEL gets removed as dead code since its result is unused.

I think the tests are not running at all in CI. If I make a checked x86 build, I can still hit the assert with MinOpts=0/TieredCompilation=0.

jakobbotsch commented:
> I think the tests are not running at all in CI. If I make a checked x86 build, I can still hit the assert with MinOpts=0/TieredCompilation=0.

It's weird, I cannot find the tests in the Helix "Tests" tab but I do see them running if I dig through the logs manually, for example: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-a295e0adf2d646a5af/JIT.HardwareIntrinsics.X86.Avx1/1/console.731898b2.log?helixlogtype=result

BTW, I would expect this to be a problem on both x86 and x64 (not sure why the original issue is marked arch-x86, the assert is coming from a 64-bit build there)

jakobbotsch reviewed (approved):
This looks fine to me. You could consider just checking `info.compHasNextCallRetAddr` at the beginning of `impIntrinsic` and early outing there, which would allow you to delete the other checks. We will never see `compHasNextCallRetAddr` in contexts where `mustExpand` will be true (and fundamentally cannot support that).

kunalspathak commented:
> The test only fails in debug codegen. In checked/release I expect that the GT_LABEL gets removed as dead code since its result is unused.

Why is debug and checked code paths different? If the test fails in debug I would expect it to fail in checked too, right?

jakobbotsch commented:
> > The test only fails in debug codegen. In checked/release I expect that the GT_LABEL gets removed as dead code since its result is unused.
> 
> Why is debug and checked code paths different? If the test fails in debug I would expect it to fail in checked too, right?

In debug builds System.Private.CoreLib.dll has the "debug bit" set, I think that causes the dynamic methods created to get debug codegen so no JIT optimizations.

In checked builds System.Private.CoreLib.dll does not have the "debug bit" set so I think the dynamic methods get optimized. That was my guess at least, I'm not sure given that @TIHan said it reproduces on checked x86 builds above.

kunalspathak commented on src/coreclr/jit/codegenlinear.cpp:
```diff
@@ -847,6 +847,13 @@ void CodeGen::genCodeForBBlist()
         }
 #endif // defined(DEBUG) && defined(USING_VARIABLE_LIVE_RANGE)
 
+#ifdef DEBUG
```
Won't this assert always be `true` given that at the beginning of this method we do:

```c++
bool needLabel = (block->bbFlags & BBF_HAS_LABEL) != 0;
...
...
if (needLabel)
{
    // Mark a label and update the current set of live GC refs

    block->bbEmitCookie = GetEmitter()->emitAddLabel(gcInfo.gcVarPtrSetCur, gcInfo.gcRegGCrefSetCur,
                                                     gcInfo.gcRegByrefSetCur, false DEBUG_ARG(block));
}
```

TIHan replied:
Unless something else could modify it, which I don't think that is happening. I put it there for a sanity check.

TIHan commented:
> You could consider just checking info.compHasNextCallRetAddr at the beginning of impIntrinsic

I thought about this, I'll see if I can cleanly do it.

TIHan committed:
Conslidate compHasNextCallRetAddr

TIHan committed:
Remove extra assert

TIHan commented:
> In checked builds System.Private.CoreLib.dll does not have the "debug bit" set so I think the dynamic methods get optimized. That was my guess at least, I'm not sure given that @TIHan said it reproduces on checked x86 builds above.

I think this is right. I double checked to see if it appears in a checked build and it does not; I was incorrect.

TIHan committed:
Fixing some logic

TIHan committed:
Remove opt check

TIHan committed:
Reverting back to original nextcallretaddr logic

TIHan committed:
Fix formatting

TIHan committed:
Fix formatting

TIHan commented:
@jakobbotsch I tried to add this logic at the start of `impIntrinsic` but I couldn't seem to get it to pass CI, so I went back to what I had before. In a separate PR, I think we could add similar logic at the beginning of `impIntrinsic` - but for now, this solves a few existing issues with reflection and hwintrinsics. I made an issue: https://github.com/dotnet/runtime/issues/72972
